### PR TITLE
Dataset class and tests for LLAMA3; support for streaming, parquet

### PR DESCRIPTION
**Pull request by savitha-eng**
### Description

This PR adds a genomic dataset module for training Llama3 models on nucleotide sequences, following the ESM2 native TE pattern (much of the code is similar to the ESM2 dataset). 

**Key features:**
- **Streaming Parquet datasets**: Efficient loading of large genomic datasets using HuggingFace `datasets` library with `streaming=True` to avoid loading entire datasets into memory
- **Windowing/Strided sampling**: Automatic creation of overlapping windows from long genomic sequences using the tokenizer's built-in `return_overflowing_tokens=True` parameter
- **Shuffle buffer**: Large shuffle buffer (500K samples by default) for better randomization during streaming
- **Distributed training support**: Built-in support for multi-GPU training with `split_dataset_by_node`
- **Causal LM collation**: Uses `DataCollatorForLanguageModeling` with `mlm=False` for next-token prediction

**Implementation details:**
- `create_tokenized_dataset()`: Loads Parquet data, handles dataset splits, applies tokenization with windowing
- `create_bshd_dataloader()`: Creates PyTorch DataLoader with appropriate sampler and collator
- Supports both streaming and non-streaming modes
- Supports both lazy and eager tokenization

**Testing:**
- 8 dataset tests covering windowing, streaming, lazy tokenization, and batch structure
- Mock data fixtures in `conftest.py` for CI/CD compatibility
- Note that the tests focus on testing the single node behavior/functionality of the dataloader, distributed dataset tests following the esm2 pattern are a TODO

#### Usage

```python
from dataset import create_bshd_dataloader
from distributed_config import DistributedConfig

# Configure distributed training (defaults to single GPU if env vars not set)
distributed_config = DistributedConfig()

# Create dataloader for genomic sequences
dataloader, sampler = create_bshd_dataloader(
    distributed_config=distributed_config,
    tokenizer_path="/path/to/nucleotide_fast_tokenizer",
    load_dataset_kwargs={
        "path": "parquet",
        "data_files": "/path/to/genomic_sequences.parquet",
        "split": "train",
        "streaming": True,  # Memory-efficient streaming
    },
    micro_batch_size=4,
    max_seq_length=8192,    # Window size
    stride=200,              # Overlap between windows (200 tokens)
    buffer_size=500_000,     # Shuffle buffer size
)

# Train
for batch in dataloader:
    # batch contains: input_ids, attention_mask, labels
    # labels = input_ids (for causal LM, DataCollator handles shifting)
    outputs = model(**batch)
    loss = outputs.loss
    loss.backward()
    optimizer.step()
```

**Key parameters:**
- `max_seq_length`: Window size for genomic sequences (e.g., 8192 for full Llama3 context)
- `stride`: Overlap between consecutive windows in tokens (e.g., 200 for 200bp overlap)
- `streaming`: Set to `True` for large datasets to avoid loading everything into memory
- `buffer_size`: Shuffle buffer size for streaming mode (larger = better randomization)

### Type of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactor
- [ ] Documentation update
- [ ] Other (please describe):

### CI Pipeline Configuration

Configure CI behavior by applying the relevant labels. By default, only basic unit tests are run.

- [ciflow:skip](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:skip) - Skip all CI tests for this PR
- [ciflow:notebooks](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:notebooks) - Run Jupyter notebooks execution tests for bionemo2
- [ciflow:slow](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:slow) - Run slow single GPU integration tests marked as @pytest.mark.slow for bionemo2
- [ciflow:all](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:all) - Run all tests (unit tests, slow tests, and notebooks) for bionemo2. This label can be used to enforce running tests for all bionemo2.
- [ciflow:all-recipes](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:all-recipes) - Run tests for all recipes (under bionemo-recipes). This label can be used to enforce running tests for all recipes.

Unit tests marked as `@pytest.mark.multi_gpu` or `@pytest.mark.distributed` are not run in the PR pipeline.

For more details, see [CONTRIBUTING](CONTRIBUTING.md)

> [!NOTE]
> By default, only basic unit tests are run. Add appropriate labels to enable an additional test coverage.

#### Authorizing CI Runs

We use [copy-pr-bot](https://docs.gha-runners.nvidia.com/apps/copy-pr-bot/#automation) to manage authorization of CI
runs on NVIDIA's compute resources.

- If a pull request is opened by a trusted user and contains only trusted changes, the pull request's code will
  automatically be copied to a pull-request/ prefixed branch in the source repository (e.g. pull-request/123)
- If a pull request is opened by an untrusted user or contains untrusted changes, an NVIDIA org member must leave an
  `/ok to test` comment on the pull request to trigger CI. This will need to be done for each new commit.

### Pre-submit Checklist

- [x] I have tested these changes locally
- [x] I have updated the documentation accordingly
- [x] I have added/updated tests as needed
- [x] All existing tests pass successfully

### CODE REVIEW

File: bionemo-recipes/recipes/llama3/dataset.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    # Handle DatasetDict (extract "train" split if present)
+    if isinstance(dataset, (datasets.DatasetDict, datasets.IterableDatasetDict)):
+        if "train" in dataset:
+            dataset = dataset["train"]
+        else:
+            raise ValueError(f"Dataset has splits {list(dataset.keys())} but no 'train' split found. "
+                           "Please specify split='train' in load_dataset_kwargs or ensure your dataset has a 'train' split.")
+
+    # Normalize column names - rename 'nt_sequence' to 'sequence' if present
+    # Only do this for non-streaming datasets (streaming datasets don't have column_names attribute)
```
Comment (pstjohn) [inline bionemo-recipes/recipes/llama3/dataset.py]: won't this affect the `tokenize` function that uses the ['sequence'] key? which datasets have `nt_sequence`?

File: bionemo-recipes/recipes/llama3/tests/test_dataset.py [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+    
+    # Create a minimal dataset with one 1000bp sequence
+    sequence = "A" * 1000
+    
+    table = pa.table({
+        "sequence": [sequence],
+    })
+    
+    pq.write_table(table, parquet_path)
+    return str(parquet_path)
```
Comment (pstjohn) [inline bionemo-recipes/recipes/llama3/tests/test_dataset.py:61]: can we have a few more sequences here 😅 ? would be good to test batching

File: bionemo-recipes/recipes/llama3/tests/test_dataset.py [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+    }
+    
+    dataloader, _ = create_bshd_dataloader(
+        distributed_config=distributed_config,
+        tokenizer_path=tokenizer_path,
+        load_dataset_kwargs=load_dataset_kwargs,
+        micro_batch_size=2,
+        num_workers=0,
+        max_seq_length=500,
+        stride=100,
```
Comment (pstjohn) [inline bionemo-recipes/recipes/llama3/tests/test_dataset.py:283]: you could add a test that checks the stride stuff, right? maybe with tokens like

sequence = 1000 * "A" + 1000 * "B"
and then make sure that sequences come out the way we expect

File: bionemo-recipes/recipes/llama3/dataset.py (hunk 1 above)
Comment (savitha-eng) [reply bionemo-recipes/recipes/llama3/dataset.py]: The SQLitedataset that Bruno directed me to (which I coverted to parquet) uses nt_sequence along with contig_id and length. - not sure if this is specific to just this dataset, or whether it will be there more generally in other datasets.  but since the column is renamed to "sequence" before the call to tokenize_and_window, it shouldn't affect it.
